### ISSUE TITLE
2D Crop Layer Support

### ISSUE THREAD

**Issue by nhkrishna** [intel/mkl-dnn#290]
Hi,

Is there any primitive support for 2D Crop Layer ?. It is good to have support for 2D crop layer, because this layer used in FCN. (or) Is there any way to use existing primitives to support 2D cropping?.

Thanks,
Hari

**Comment by emfomenk** [comment]
Hi @nhkrishna,

View provides necessary functionality to implement crop. Please read the documentation [here](https://intel.github.io/mkl-dnn/group__c__api__memory.html#gab404869f8c612df54e4734e3e2cc6353), example [here](https://github.com/intel/mkl-dnn/blob/master/examples/simple_rnn.cpp#L449), and #69 issue for more details. The issue contains the most comprehensive explanation.

One way to do crop is:
```cpp
crop_view_pd(origin_mem_pd, array_offset = {0,0,1,1}, array_crop_size = {10, 20, 3, 3});
reorder_pd(crop_view_pd, cropped_mem_pd);
reoder(origin_mem, cropped_mem);
```

Also, in theory you can proceed with crop_view_pd and pass to the primitive that should work on cropped memory. In this case you even don't need to copy the cropped part outside of the original memory.

**Comment by nhkrishna** [comment]
Thanks @emfomenk  :+1:

**Comment by nhkrishna** [comment]
Hi @emfomenk ,

I have tried the cropping using above method for MKLDNN v0.13, Its working fine. But when try to use v0.17 and later i'm running into following error.
`terminate called after throwing an instance of 'mkldnn::error'                                                                                                                                                                                                                
  what():  std::exception                                                                                                                                                                                                                                                     
Abort `

Does MKLDNN v0.17 has any API related changes in version 0.17 and later ?

Thanks,
Hari

**Comment by emfomenk** [comment]
Hi @nhkrishna,

No, there are no changes wrt to API. However somewhere in Intel MKL-DNN v0.15 we introduced support for [padded memory](https://github.com/intel/mkl-dnn/blob/master/doc/design/understanding_memory_formats.md#what-if-channels-are-not-multiple-of-8-or-16), which allows using blocked layout for the sizes which are not multiple of the block size. In particular, that allowed running avx2 jitted convolution (`nChw8c`) for the cases when number of channels is not divisible by 8. That might sometimes lead to the cases when view cannot be created (in fact even in v0.13 view might have failed, but it was really hard to come up with the appropriate situation).

Here is an example:
``` c++
    auto data_md = memory::desc({3, 64, 13, 13}, f32, nChw8c);
    auto data_mpd = memory::primitive_desc(data_md, eng);

    // for let's create a windows of size 8 channels with offset 0 for channels
    view::primitive_desc(data_mpd, {3, 8, 13, 13}, {0, 0, 0, 0}); // <-- oK


    // for let's create a windows of size 8 channels with offset 4 for channels
    view::primitive_desc(data_mpd, {3, 8, 13, 13}, {0, 4, 0, 0}); // <-- exception
```

The reason is that this view cannot be expressed in a way Intel MKL-DNN describes memory. 

In other words the restriction for view is that both sizes and offsets must be multiple of block sizes if a blocked format is used (like `nChw8c` or `nChw16c`).

For this particular case you have to reorder data (`data_mpd`) to any plain layout (e.g. `nchw`) and create a view only after that.

This is a known limitation of a view, and we cannot overcome it with the current API. We have some thoughts how to fix this in the next versions (with some API changes).

**Comment by nhkrishna** [comment]
Hi @emfomenk 

Thanks for info. But in my case both source and destination memory primitives are created in `nchw` format only. But still i'm seeing the issue. I have attached the [crop2d stanalone](https://drive.google.com/open?id=1ewbLcl-rWNKOvxVd5QxJzBCWkmzTmAsw)  snnippet for which V17 is crashing and upto v16 are working. Can you please take a look into this ?

Thanks,
Hari

**Comment by emfomenk** [comment]
Hi Hari,

I tried your snippet and it worked fine for me (i removed reading and writing to file, but the remaining keep the same):
``` bash
$ ( g++ crop.cpp -lmkldnn && MKLDNN_VERBOSE=1 ./a.out )
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nchw out:f32_nchw,num:1,1x3x16x16,0.000976562
```

I used the latest master. Could you please give it a try?
If it still fails, could you please share your HW and MKL-DNN building details.

**Comment by nhkrishna** [comment]
Hi @emfomenk ,

I have tried using [v0.17.2](https://github.com/intel/mkl-dnn/releases/tag/v0.17.2) on linux platform.
And my system configuration is `Intel(R) Xeon(R) CPU E5-1650 v4 @ 3.60GHz`. 
I have followed the below steps to build MKLDNN library:
- Downloaded the v0.17.2 from MKLDNN release page.
- Downloaded the mklml_lnx_2019.0.1.20181227.tgz and placed in external folder from MKLDNN root library
-  `mkdir -p build && cd build && cmake .. && make`

I will try with MKLDNN master and get back to you.

Thanks,
Hari

**Comment by nhkrishna** [comment]
Thanks @emfomenk . With plain nchw data my code works fine in all versions. But the actual issue seems to be output channels of view primitive which is not in the multiples of 8 then i'm running into error which you mentioned earlier. After reordering the nchw8c data to nchw, resolves my issue.
